File: language/c/polymorphism/point/point.c
//PointµÄCÎÄ¼þ£ºpoint.c
 
#include "point.h"
 
#include "new.h"
 
#include "point.h"
 
#include "point.r"
 
static void * Point_ctor (void * _self, va_list * app)
{
 
	struct Point * self = _self;
 
	self->x = va_arg(*app, int);
 
	self->y = va_arg(*app, int);
 
	return self;
 
}
 
static void Point_draw (const void * _self)
{
 
	const struct Point * self = _self;
 
	printf("draw (%d,%d)", self -> x, self -> y);
}
 
static const struct Base _Point = 
{
	sizeof(struct Point), Point_ctor, 0, Point_draw
};
 
const void * Point = &_Point;

File: language/c/polymorphism/point/point.r
//Point内部头文件(外面看不到)：point.r
 
#ifndef POINT_R
 
#define POINT_R
 
struct Point
{
    const void * base; //继承，基类指针，放在第一个位置，const是防止修改
    int x, y;        //坐标
};
 
#endif